### PR TITLE
🐛 Lookup DNS IP in Service

### LINKED ISSUE #2365: bug: Retrieve the workspace-related CoreDNS IP from the Service resource instead of using DNS lookup.
### Describe the bug

Currently, when mutating a Deployment to override the DNS server, the Syncer tries to guess the IP of the CoreDNS service by issuing a DNSLookup on the internal service dns name.
This prevents using the Syncer process outside of the Synctarget cluster (for local debugging purposes for example).
A better and more robust way is to get the `clusterIP` on the Service resource.

### Steps To Reproduce

1. Run the Syncer in it own local process
2. Observe that, when trying to sync a deployment, you would be the error:
```
E1115 18:44:16.121384  418586 spec_controller.go:295] kcp-workload-syncer-spec failed to sync {{"apps" "v1" "deployments"} "root:org:user|default/httpecho"}, err: failed to get DNS nameserver IP address: lookup kcp-dns-us-west1-1mwhl3gm-x0gx4iqu.kcp-syncer-us-west1-1mwhl3gm.svc.cluster.local: no such host
```

### Expected Behaviour

We should still be able to run the syncer locally.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by lionelvillard**
## Summary

- Get the DNS IP address from the Service object.
- Rename BoundWorkspace to WorkspaceBound

## Related issue(s)

Fixes #2365

### CODE REVIEW

File: pkg/syncer/spec/mutators/deployment.go [hunk 1]
```diff
@@ -61,18 +55,18 @@ func (dm *DeploymentMutator) GVR() schema.GroupVersionResource {
 	}
 }
 
-func NewDeploymentMutator(upstreamURL *url.URL, secretLister ListSecretFunc, syncTargetLogicalClusterName logicalcluster.Name,
+func NewDeploymentMutator(upstreamURL *url.URL, secretLister ListSecretFunc, serviceLister listerscorev1.ServiceLister,
```
Comment (davidfestal) [inline pkg/syncer/spec/mutators/deployment.go:58]: is it possible to use a `ListSecretFunc` following the same pattern as what was used for secrets ?

File: pkg/syncer/spec/mutators/deployment_test.go [hunk 2]
```diff
@@ -839,9 +840,14 @@ func TestDeploymentMutate(t *testing.T) {
 
 				workspace := logicalcluster.New("root:default:testing")
 
-				dm := NewDeploymentMutator(upstreamURL, secretLister, workspace, "syncTargetUID", "syncTargetName", "dnsNamespace")
+				serviceIndexer := cache.NewIndexer(cache.LegacyMetaNamespaceKeyFunc, cache.Indexers{})
```
Comment (davidfestal) [inline pkg/syncer/spec/mutators/deployment_test.go:843]: Coujld we do the same as for the `secretLister` ? (cf previous comment)

**Comment by lionelvillard** [comment]
/hold

testing with David fixes.

**Comment by lionelvillard** [comment]
Got this error:

```
virtualworkspace_test.go:558: Patch luckyluke via virtual workspace to report in status that joe is in prison
I1116 18:50:29.484716   50955 spec_process.go:350] "Updated upstream resource with syncer finalizer" syncTarget.workspace="root:e2e-org-dgpv2:e2e-workspace-bjvgl" syncTarget.name="wildwest-356231" syncTarget.key="2oMmwW6bPkUX2W07XmZgfjISikTMIbUBI6fpoK" reconciler="kcp-workload-syncer-spec" key="root:e2e-org-dgpv2:e2e-workspace-bjvgl|kcp-system/kube-root-ca.crt" gvr="/v1, Resource=configmaps"
I1116 18:50:29.485041   50955 spec_controller.go:288] "processing key" syncTarget.workspace="root:e2e-org-dgpv2:e2e-workspace-bjvgl" syncTarget.name="wildwest-356231" syncTarget.key="2oMmwW6bPkUX2W07XmZgfjISikTMIbUBI6fpoK" reconciler="kcp-workload-syncer-spec" key="root:e2e-org-dgpv2:e2e-workspace-bjvgl|kcp-system/default-token-flt7f" gvr="/v1, Resource=secrets"
I1116 18:50:29.487713   50955 spec_controller.go:246] "queueing GVR" syncTarget.workspace="root:e2e-org-dgpv2:e2e-workspace-bjvgl" syncTarget.name="wildwest-356231" syncTarget.key="2oMmwW6bPkUX2W07XmZgfjISikTMIbUBI6fpoK" reconciler="kcp-workload-syncer-spec" key="root:e2e-org-dgpv2:e2e-workspace-bjvgl|kcp-system/kube-root-ca.crt" gvr="/v1, Resource=configmaps"
    virtualworkspace_test.go:560: 
        	Error Trace:	virtualworkspace_test.go:560
        	            				virtualworkspace_test.go:1460
        	Error:      	Received unexpected error:
        	            	Internal error occurred: tried to write resource cowboys:ec4f85f4e25edbc3278635fb61ec25f11881ff964f621e96b30c8fe3b9836426/status(default/luckyluke) though it is not owning it (syncer finalizer doesn't exist)
```

**Comment by davidfestal** [comment]
> Got this error:
> 
> ```
> virtualworkspace_test.go:558: Patch luckyluke via virtual workspace to report in status that joe is in prison
> I1116 18:50:29.484716   50955 spec_process.go:350] "Updated upstream resource with syncer finalizer" syncTarget.workspace="root:e2e-org-dgpv2:e2e-workspace-bjvgl" syncTarget.name="wildwest-356231" syncTarget.key="2oMmwW6bPkUX2W07XmZgfjISikTMIbUBI6fpoK" reconciler="kcp-workload-syncer-spec" key="root:e2e-org-dgpv2:e2e-workspace-bjvgl|kcp-system/kube-root-ca.crt" gvr="/v1, Resource=configmaps"
> I1116 18:50:29.485041   50955 spec_controller.go:288] "processing key" syncTarget.workspace="root:e2e-org-dgpv2:e2e-workspace-bjvgl" syncTarget.name="wildwest-356231" syncTarget.key="2oMmwW6bPkUX2W07XmZgfjISikTMIbUBI6fpoK" reconciler="kcp-workload-syncer-spec" key="root:e2e-org-dgpv2:e2e-workspace-bjvgl|kcp-system/default-token-flt7f" gvr="/v1, Resource=secrets"
> I1116 18:50:29.487713   50955 spec_controller.go:246] "queueing GVR" syncTarget.workspace="root:e2e-org-dgpv2:e2e-workspace-bjvgl" syncTarget.name="wildwest-356231" syncTarget.key="2oMmwW6bPkUX2W07XmZgfjISikTMIbUBI6fpoK" reconciler="kcp-workload-syncer-spec" key="root:e2e-org-dgpv2:e2e-workspace-bjvgl|kcp-system/kube-root-ca.crt" gvr="/v1, Resource=configmaps"
>     virtualworkspace_test.go:560: 
>         	Error Trace:	virtualworkspace_test.go:560
>         	            				virtualworkspace_test.go:1460
>         	Error:      	Received unexpected error:
>         	            	Internal error occurred: tried to write resource cowboys:ec4f85f4e25edbc3278635fb61ec25f11881ff964f621e96b30c8fe3b9836426/status(default/luckyluke) though it is not owning it (syncer finalizer doesn't exist)
> ```
> 
> FYI @davidfestal @jmprusi

This might simply be because, in this test case, we don't wait for the syncer to add the finalizer before trying to patch the resource. So if the syncer is slower to add the finalizer, this would fail here.

File: pkg/syncer/spec/mutators/deployment.go (hunk 1 above)
Comment (davidfestal) [reply pkg/syncer/spec/mutators/deployment.go:58]: Got it

**Pull request merged**